# 01 - Data Understanding

## Objective

Notebook ini digunakan untuk memahami struktur awal dataset Customer 360
sebelum melakukan cleaning, standardization, atau entity resolution.

## Goals

1. Load dataset
2. Memahami struktur dataset
3. Mengetahui jumlah rows dan columns
4. Memeriksa nama kolom
5. Memeriksa data types
6. Memeriksa missing values
7. Memeriksa unique values
8. Memeriksa exact duplicates
9. Melihat sample records
10. Mengidentifikasi field yang berpotensi digunakan untuk entity resolution
11. Menentukan pertanyaan yang perlu dijawab sebelum matching

> Important:
> Tidak ada cleaning atau transformasi permanen pada notebook ini.

In [2]:
import pandas as pd
import numpy as np
from pathlib import Path

In [6]:
DATA_PATH = Path(r"C:\Users\User\Documents\Maganghub 2026\Bulan 1\Tes duplikasi\data\raw\crm_50000_customers_dirty_v3.csv")

df = pd.read_csv(DATA_PATH)

In [7]:
# Baca data
print(f"File: {DATA_PATH}")
print(f"Exists: {DATA_PATH.exists()}")

File: C:\Users\User\Documents\Maganghub 2026\Bulan 1\Tes duplikasi\data\raw\crm_50000_customers_dirty_v3.csv
Exists: True


In [8]:
# Ukuran
n_rows, n_cols = df.shape

print(f"Rows    : {n_rows:,}")
print(f"Columns : {n_cols}")

# Nama Kolom
print("Columns:")
for i, col in enumerate(df.columns, start=1):
    print(f"{i:2}. {col}")

# Struktur
df.info

Rows    : 50,000
Columns : 14
Columns:
 1. customer_id
 2. first_name
 3. last_name
 4. email
 5. phone_number
 6. gender
 7. dob
 8. signup_date
 9. address
10. city
11. state
12. country
13. device_id(s)
14. source


<bound method DataFrame.info of                                 customer_id first_name    last_name  \
0      0d244537-8164-4b84-bcf2-0e43766f3221      mARIA          day   
1      66c5800b-ad5c-4fac-86df-16472d82f02d   KKeevvin        Cantu   
2      e7852eee-0a28-46dc-aa34-445a02530746     Ronald       Daniel   
3      14f0e67d-e60e-4989-88cf-5be50d515a69  Katherine        Smith   
4      1ce2f061-345c-4121-a4ba-32963cf9a15d    Brandon       Glover   
...                                     ...        ...          ...   
49995  fa5edcf0-32a4-45b5-a28c-c0696c1fd386      amber   Henderson    
49996  e19d4835-e3d2-4358-b8ee-aa4297aecda1       Toni        Moore   
49997  b4749ea0-7d83-42ef-8506-548f9f31fb1f       John        Davis   
49998  9bf5e55e-de53-462d-9b8b-52e2aeabc48f      Tracy     Sandoval   
49999  3d2cc98d-7c76-4cf3-9b04-fb0f909513d7       Mark  Fitzpatrick   

                               email           phone_number gender  \
0           maria.day697@hotmail.com       44

In [9]:
# Data Type
dtype_summary = pd.DataFrame({
    "column": df.columns,
    "dtype": df.dtypes.astype(str).values
})

dtype_summary

# Sampel isi data
df.sample(10, random_state=42)


,customer_id,first_name,last_name,email,phone_number,gender,dob,signup_date,address,city,state,country,device_id(s),source
33553,14b19db2-25f2-4c42-afac-822b0412f55c,Nicole,Potts,nicole.potts659@gmail.com,001-640-729-1553x3879,F,1971-10-21,2015-05-10,245 Myers Union Apt. 852,Schmidtview,New Jersey,Bosnia and Herzegovina,3fbc6171-6d49-4080-a482-c4c99af7e158,web
9427,11eae6a0-c04d-4e90-aaeb-c312cdd85f4c,Jessica*,HHaanncock,jessica.hancock630@gmail.com,230.679.9769,F,1989-12-16,2020-12-01,686 Julia Union Apt. 529,West Brian,Oregon,Maldives,64bd5dca-1628-464d-a433-1a27dc42bfaa,app
199,c534fb8b-d99a-444c-943b-1ec88d67be00,Kimberly,Cox!,kimberly.cox676@yahoo.com,087-265-6604,F,2005-07-04,2015-05-22,442 Katrina Drives,North Andrea,Minnesota,Svalbard & Jan Mayen Islands,bc7bf030-33da-4093-93d8-404ab675f0ef,web
12447,bf14ab2c-de04-4141-894d-f9538c71a8f6,Danielle,Johnson,danielle.johnson048@yahoo.com,553.080.1979,M,1971-04-13,2019-08-12,59419 Williams Squares Apt. 528,New Brandonton,Oklahoma,Belgium,24c21a12-7e3a-4a78-a143-e293eb669179,app
39489,96ba3e93-5bd6-44d0-8413-96af38b371a0,Darren,Rogers,darren.rogers723@yahoo.com,(264)226-4218,F,1998-03-15,2020-05-31,5919 Coleman Crest,Glendaland,Maryland,Australia,63a197f7-0092-40e2-a251-d75b81e4efcd,app
42724,98b84fac-290f-4e0d-bed0-7877e31236ff,Diane,Wilkins,diane.wilkins421@yahoo.com,001-272-370-9902x682,F,1966-02-21,2024-03-06,052 Patel Islands,New Brad,South Dakota,Jordan,6802f7bd-958e-4ab2-ad12-e350230ec1e2,web
10822,153e1fef-c8db-43cc-b92c-051cc6e7c299,Valerie,Walker,valerie.walker476@yahoo.com,+1-897-770-1091x7231,M,1999-09-26,2015-04-17,5137 Chapman Falls Apt. 415,Angelachester,Oklahoma,Japan,a1b27a41-d546-46ac-bc6f-25202acde3ea,referral
49498,a8ae12da-f250-4677-9305-09986b4bc8a6,Carla,Acevedo,carla.acevedo938@hotmail.com,+1-370-645-8796x99924,O,1996-12-27,2017-06-20,801 Jones Lights,New Timothybury,New Jersey,Djibouti,4ee56951-9555-4da6-9083-8f4ab527c143,app
4144,2123f54a-9d6e-43f9-a29b-1e7e6219fa05,Dylan,Manning,dylan.manning784@yahoo.com,+1-413-997-4858x0456,M,1999-05-06,2019-04-01,77227 Eric Isle Apt. 901,New Terristad,Alaska,Israel,6b87c436-ae2e-4958-ae3d-6f8ff7b3a2dd,web
36958,dbd0e933-8627-42c8-a0e3-028813cc2447,LINDA,ANDRADE,linda.andrade601@gmail.com,448-698-2727x632,F,1972-01-24,2018-08-22,7395 Miller Neck,Port Janiceshire,North Dakota,Kiribati,8cd42ba2-adf1-4e88-8306-c25dc44784c4,web


In [10]:
# Cek Missing
missing_summary = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_percentage": df.isna().mean() * 100
})

missing_summary = missing_summary.sort_values(
    "missing_percentage",
    ascending=False
)

missing_summary.round(2)

,missing_count,missing_percentage
email,1040,2.08
customer_id,0,0.00
first_name,0,0.00
last_name,0,0.00
phone_number,0,0.00
gender,0,0.00
dob,0,0.00
signup_date,0,0.00
address,0,0.00
city,0,0.00


In [11]:
# Deteksi Missing values
df.isna()

,customer_id,first_name,last_name,email,phone_number,gender,dob,signup_date,address,city,state,country,device_id(s),source
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,False,False,False,False,False,False,False,False,False,False,False,False,False,False
49996,False,False,False,False,False,False,False,False,False,False,False,False,False,False
49997,False,False,False,False,False,False,False,False,False,False,False,False,False,False
49998,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [12]:
# Cek Uniqe values
unique_summary = pd.DataFrame({
    "unique_count": df.nunique(dropna=True),
    "unique_percentage": (
        df.nunique(dropna=True) / len(df) * 100
    )
})

unique_summary = unique_summary.sort_values(
    "unique_count",
    ascending=False
)

unique_summary.round(2)

,unique_count,unique_percentage
customer_id,48200,96.40
address,48200,96.40
device_id(s),48200,96.40
phone_number,46777,93.55
email,46363,92.73
city,24534,49.07
dob,17733,35.47
last_name,7228,14.46
first_name,5429,10.86
signup_date,4383,8.77


Persiapan ke Duplicate

In [13]:
#exact
exact_duplicate_count = df.duplicated().sum()

print(f"Exact duplicate rows: {exact_duplicate_count:,}")

exact_duplicate_percentage = (
    exact_duplicate_count / len(df) * 100
)

print(f"Percentage: {exact_duplicate_percentage:.2f}%")

Exact duplicate rows: 1,021
Percentage: 2.04%


In [14]:
df[df.duplicated(keep=False)].head(20)

,customer_id,first_name,last_name,email,phone_number,gender,dob,signup_date,address,city,state,country,device_id(s),source
4,1ce2f061-345c-4121-a4ba-32963cf9a15d,Brandon,Glover,brandon.glover804@hotmail.com,489.732.5436,M,1968-06-25,2022-01-09,55037 Denise Crest Apt. 565,Port Amber,Oregon,Latvia,751d4410-3d73-4f4c-b432-10a0213b925f,referral
5,6ddc1772-f210-4111-87c9-c2ec73776c37,Toni,Li,toni.li092@gmail.com,(522)382-8559,F,1981-09-30,2023-11-25,1230 Padilla Parks Suite 647,North Melvintown,Kentucky,Uganda,f8013534-0730-4d00-8120-445e02ebed1f,app
47,ac66a771-49b4-4aac-837b-9e7321e90daf,Lori,King,lori.king997@gmail.com,+1-246-907-6868,F,1959-05-15,2014-10-12,666 Gray Fort Apt. 650,Rachelmouth,Rhode Island,Uganda,61b3d014-8c49-4573-935e-d46d33fc8467,referral
59,b9b77c50-6a92-4a4c-a0bb-4dd450c3ce18,Philip,Obrien,philip.obrien138@hotmail.com,129-908-6602x0924,M,1978-02-16,2021-05-28,51734 Nathan Junction,Maymouth,Texas,Namibia,2d21e193-03ba-4da9-8967-9d931438e8e2,web
113,c6b4046e-9768-45d1-bd10-e75f47611328,David,Gray,david.gray928@hotmail.com,862.238.1108x954,F,1997-01-26,2019-09-24,350 Norman Mews Apt. 305,Justinhaven,Pennsylvania,Tokelau,ff00b523-ae04-45ab-8aed-3fe3f79e9d49,referral
159,977a1f54-c82b-42ad-87d3-fa383e17ce63,David,Mcmahon,david.mcmahon960@gmail.com,(416)321-3395,M,1994-11-24,2023-04-28,9531 Bryan Green,Port Alan,Georgia,Cote d'Ivoire,882aa5fe-f81a-44df-b908-ec9504a01124,web
160,a59c953e-11f3-444d-a991-5c8219307f14,Daniel,Bishop,daniel.bishop342@yahoo.com,(213)928-7234x3832,F,1980-12-22,2016-03-30,243 Janet Hills Apt. 905,Lake Heathershire,Ohio,Libyan Arab Jamahiriya,eee90259-6af3-43d5-b155-8346ce75eedb,app
164,9e22b517-a832-4a5d-b0df-9e3abd7c93de,Jacqueline,Chandler,jacqueline.chandler045@yahoo.com,(971)553-4249x49578,M,1995-11-12,2025-06-30,125 Robinson Villages,Timothymouth,Nevada,Japan,90e7f8c1-f18c-4665-a05b-5433524cf28e,app
180,ef21c862-9997-4889-b718-117527ba4c42,John,Martin,john.martin468@hotmail.com,001-982-573-8587x0008,F,1980-11-06,2014-09-23,373 Martin Throughway Suite 576,Johnberg,Vermont,Senegal,af1c56fc-ca8b-4605-bfc3-ef179cf43925,app
183,0261ca6d-1cce-46ce-b7bd-199d2a40bd40,Wanda,Pacheco,wanda.pacheco887@gmail.com,+1-626-285-4188x0598,M,1985-01-14,2016-12-09,49136 Thomas Glen,Warnershire,Connecticut,Grenada,3ad887ac-c081-4931-857d-252b359f5bb3,referral


Konsepnya:
Exact duplicate ≠ entity duplicate.

In [15]:
if "customer_id" in df.columns:
    duplicated_customer_id = df["customer_id"].duplicated().sum()

    print(f"Duplicated customer_id: {duplicated_customer_id:,}")

if "customer_id" in df.columns:
    print(
        f"Unique customer_id: "
        f"{df['customer_id'].nunique(dropna=True):,}"
    )

Duplicated customer_id: 1,800
Unique customer_id: 48,200


In [19]:
cardinality = pd.DataFrame({
    "column": df.columns,
    "rows": len(df),
    "unique": [
        df[col].nunique(dropna=True)
        for col in df.columns
    ],
    "unique_ratio": [
        df[col].nunique(dropna=True) / len(df)
        for col in df.columns
    ]
})

cardinality = cardinality.sort_values(
    "unique_ratio",
    ascending=False
)

cardinality.round(4)

,column,rows,unique,unique_ratio
0,customer_id,50000,48200,0.9640
8,address,50000,48200,0.9640
12,device_id(s),50000,48200,0.9640
4,phone_number,50000,46777,0.9355
3,email,50000,46363,0.9273
9,city,50000,24534,0.4907
6,dob,50000,17733,0.3547
2,last_name,50000,7228,0.1446
1,first_name,50000,5429,0.1086
7,signup_date,50000,4383,0.0877


In [16]:
for col in df.columns:
    unique_count = df[col].nunique(dropna=True)

    if unique_count <= 20:
        print(f"\n{'=' * 50}")
        print(f"{col}")
        print(f"{'=' * 50}")
        print(df[col].value_counts(dropna=False))


gender
gender
F    23333
M    22880
O     3787
Name: count, dtype: int64

source
source
referral    16768
app         16692
web         16540
Name: count, dtype: int64


## Initial Entity Resolution Field Assessment

Pada tahap ini kita belum menentukan algoritma matching.

Field akan dikategorikan berdasarkan karakteristik awal:

- Potential identifier
- Potential strong matching field
- Potential supporting field
- Low-discriminative field
- Field yang perlu investigasi lebih lanjut

Penentuan final dilakukan setelah data quality dan duplicate pattern analysis.

In [17]:
candidate_fields = [
    col for col in df.columns
    if any(keyword in col.lower() for keyword in [
        "id",
        "email",
        "phone",
        "name",
        "address",
        "dob",
        "date"
    ])
]

candidate_fields

['customer_id',
 'first_name',
 'last_name',
 'email',
 'phone_number',
 'dob',
 'signup_date',
 'address',
 'device_id(s)']

In [20]:
candidate_profile = cardinality[
    cardinality["column"].isin(candidate_fields)
].copy()

candidate_profile.round(4)

,column,rows,unique,unique_ratio
0,customer_id,50000,48200,0.9640
8,address,50000,48200,0.9640
12,device_id(s),50000,48200,0.9640
4,phone_number,50000,46777,0.9355
3,email,50000,46363,0.9273
6,dob,50000,17733,0.3547
2,last_name,50000,7228,0.1446
1,first_name,50000,5429,0.1086
7,signup_date,50000,4383,0.0877


# Questions Before Matching

Sebelum masuk ke exact matching atau fuzzy matching, hasil profiling awal menjawab:

1. **Apakah `customer_id` benar-benar unik?** Tidak. Dataset memiliki 48.200 nilai unik dari 50.000 baris, sehingga terdapat 1.800 baris dengan `customer_id` yang berulang. Tidak ada nilai missing.
2. **Apakah terdapat exact duplicate records?** Ya, terdapat 1.021 baris exact duplicate atau sekitar 2,04% dari seluruh baris.
3. **Seberapa banyak missing value pada setiap field?** Missing value hanya ditemukan pada `email`, yaitu 1.040 baris atau 2,08%. Field lainnya tidak memiliki missing value berdasarkan `isna()`.
4. **Apakah email cukup lengkap dan cukup unik?** Cukup lengkap karena 48.960 dari 50.000 baris terisi (97,92%). Email memiliki 46.363 nilai unik. Seluruh email yang terisi lolos pemeriksaan format sederhana, tetapi keunikan belum berarti email bebas dari variasi huruf besar-kecil atau kesalahan semantik.
5. **Apakah phone number cukup lengkap dan cukup unik?** Lengkap karena tidak ada missing value. Terdapat 46.777 nilai unik (93,55%), tetapi panjang angka setelah karakter non-digit dihapus bervariasi antara 10 sampai 18 digit. Ini menunjukkan format nomor telepon belum konsisten dan perlu standardisasi sebelum matching.
6. **Seberapa banyak variasi pada nama?** Gabungan `first_name` dan `last_name` memiliki 42.585 kombinasi mentah dan 40.254 kombinasi setelah normalisasi sederhana. Selisih 2.331 menunjukkan adanya variasi penulisan atau tanda baca/spasi yang perlu diperiksa.
7. **Apakah terdapat indikasi typo atau karakter yang tidak konsisten?** Ada indikasi variasi pada nama dan telepon berdasarkan perubahan cardinality dan panjang nomor. Typo belum dapat dipastikan tanpa aturan validasi atau pemeriksaan pasangan record.
8. **Apakah format email konsisten?** Seluruh email terisi lolos regex format dasar `nama@domain.tld`. Konsistensi yang lebih dalam, seperti domain, case, spasi tersembunyi, atau typo domain, masih perlu diperiksa.
9. **Apakah format nomor telepon konsisten?** Tidak. Panjang digit bervariasi dari 10 hingga 18 digit dan jumlah format mentah berbeda dari jumlah format setelah normalisasi belum dianalisis per pola negara.
10. **Apakah address memiliki variasi format yang signifikan?** `address` memiliki 48.200 nilai unik, sama dengan cardinality `customer_id` dan `device_id(s)`. Normalisasi spasi sederhana tidak mengurangi jumlah nilai unik, sehingga variasi alamat tidak terbukti hanya disebabkan oleh spasi. Variasi tanda baca, singkatan, dan urutan komponen masih perlu dianalisis.
11. **Apakah dob memiliki format yang konsisten?** Semua 50.000 nilai berhasil diparse oleh `pandas.to_datetime`, tetapi format teks asli belum diuji secara eksplisit. `dob` memiliki 17.733 nilai unik, jadi konsistensi format dan kemungkinan tanggal yang sama dengan representasi berbeda masih perlu diperiksa.
12. **Apakah terdapat field yang terlalu banyak missing sehingga tidak reliable?** Tidak ada field dengan missing sangat tinggi. Hanya `email` yang missing 2,08%, sehingga email tetap usable tetapi perlu perlakuan khusus untuk baris kosong.
13. **Apakah terdapat field yang sangat discriminative?** `customer_id`, `address`, dan `device_id(s)` masing-masing memiliki 48.200 nilai unik (96,40%). `phone_number` memiliki 46.777 (93,55%) dan `email` 46.363 (92,73%). Field-field ini paling discriminative secara cardinality, tetapi belum otomatis dapat dianggap sebagai identifier yang benar.
14. **Apakah duplicate kemungkinan terjadi karena typo, formatting, missing value, atau kombinasi beberapa field?** Profiling membuktikan adanya exact duplicate, `customer_id` berulang, missing email, serta variasi nama dan telepon. Penyebab setiap duplicate belum dapat dipastikan; kemungkinan yang paling masuk akal adalah kombinasi exact duplication, variasi formatting, dan perubahan/ketidaklengkapan field.
15. **Apakah dataset menyediakan ground truth untuk mengevaluasi entity resolution?** Belum terlihat. Kolom yang tersedia tidak menunjukkan label pasangan duplicate atau `master_customer_id`. Hal ini perlu dikonfirmasi dari dokumentasi sumber data sebelum evaluasi supervised dilakukan.

Kesimpulan ini masih berada pada tahap profiling. Belum dilakukan cleaning, standardization, blocking, fuzzy matching, atau entity resolution.

# Initial Findings

## Dataset Structure

- Number of rows: **50,000**
- Number of columns: **14**
- Data types: seluruh kolom terbaca sebagai `str`; kolom tanggal (`dob` dan `signup_date`) belum dikonversi ke tipe datetime.

## Data Quality

- Missing values: hanya `email` yang memiliki missing value, yaitu **1.040 baris (2,08%)**. Kolom lain tidak memiliki missing value.
- Exact duplicate rows: **1.021 (2,04%)**
- High-cardinality fields: `customer_id`, `address`, dan `device_id(s)` masing-masing **48.200 unik (96,40%)**; `phone_number` **46.777 unik (93,55%)**; `email` **46.363 unik (92,73%)**.
- Low-cardinality fields: `gender` dan `source` masing-masing memiliki **3 nilai unik**; `state` memiliki **50**; `country` memiliki **243**. Field ini kurang cocok sebagai kunci tunggal untuk entity resolution.

## Potential Entity Resolution Fields

Berdasarkan profiling awal, field yang perlu diperiksa lebih lanjut:

- `email`: cukup lengkap dan cukup unik, tetapi 1.040 nilai missing.
- `phone_number`: cukup unik, tetapi memiliki panjang digit 10 sampai 18 sehingga perlu standardisasi.
- Kombinasi `first_name`, `last_name`, `dob`, dan `address`: dapat menjadi supporting fields karena variasi nama dan alamat perlu dibandingkan bersama.

## Important Observations

- `customer_id` tidak unik: 48.200 nilai unik untuk 50.000 baris, dengan 1.800 baris berulang.
- Terdapat 1.021 exact duplicate rows. Exact duplicate perlu dipisahkan dari entity duplicate karena record yang sama persis belum menjelaskan seluruh pola identitas pelanggan.
- `email` adalah satu-satunya field dengan missing value, sebesar 2,08%.
- `phone_number` memiliki format yang belum seragam berdasarkan panjang digit 10 sampai 18.
- Normalisasi sederhana pada gabungan nama mengurangi jumlah nilai unik dari 42.585 menjadi 40.254; ini mengindikasikan variasi penulisan yang perlu ditelusuri.
- Seluruh nilai `dob` dan `signup_date` berhasil diparse, tetapi format string asli belum dinilai secara eksplisit.

## Open Questions

1. Apakah pengulangan `customer_id` menunjukkan satu customer dengan banyak record, atau kesalahan pembuatan identifier?
2. Apakah variasi `phone_number`, nama, dan email berasal dari sumber data yang berbeda atau dari perubahan data pelanggan?
3. Apakah tersedia ground truth seperti `master_customer_id` atau label pasangan duplicate untuk mengukur precision dan recall matching?

## Next Step

Tahap berikutnya adalah Data Quality Assessment dan Duplicate Pattern Analysis.

Belum dilakukan:
- cleaning permanen
- standardization
- fuzzy matching
- blocking
- entity resolution